### ISSUE TITLE
Please add support for elixir 

### ISSUE THREAD

**Issue by mandarvaze** [jacktasia/dumb-jump#142]
`alchemist.el` which is special mode handler for elixir, already supports "go to definition" and "jump back"

**Comment by jacktasia** [comment]
Thanks for opening this issue. I'll look into adding this as soon as I have time.

**Comment by mandarvaze** [comment]
Thanks.

So when can I get these changes? 
I installed dumb-jump via MELPA

**Comment by jacktasia** [comment]
@mandarvaze The changes are on MELPA now (it can take up to 6 hours for MELPA to pick up changes), but just make sure you refresh your package list. Please let me know if you run into any issues. Thanks!

**Comment by mandarvaze** [comment]
I now have `20170810.39` version of dumb-jump
"go to definition" seems "hit-and-miss"

e.g. In 3 tries, one of them worked, other could not find the `def` defined in the same file
In the third case, it did not work, but got an error like "Took over 2s to find ... please install ag or rg.." But I don't think `ag` not accessible since `C-M-g` and `C-M-p` works well for `python` project.

case where `def` could not be found was case when function was called without `()`, since it was called via pipeline operator, where output of previous operation, is passed as first argument to the function. If the function has only one param, we do not need `()`

**Comment by jacktasia** [comment]
@mandarvaze Thanks for the report! Would you be willing to make a full example of the two not working cases? This will make it way easier for me to fix since I don't use elixir myself. Thanks!

cc @rockwood any ideas?

**Comment by rockwood** [comment]
@mandarvaze Thanks for the heads up. 

I've only tried with `ag` myself, and that seems to be working pretty well. Would you be able to put together an example repo that I could take a look at?

**Comment by mandarvaze** [comment]
@rockwood  I have `ag` installed on my machine, so I assume dumb-jump is already using that (when it works for say python) Is there a way to find out for certain ?

@jacktasia Here is a code file.

```elixir
defmodule Discuss.AuthController do
  use Discuss.Web, :controller
  plug Ueberauth
  alias Discuss.User

  def callback(%{assigns: %{ueberauth_auth: auth}} = conn, params) do
    user_params = %{token: auth.credentials.token,
                    email: auth.info.email,
                    provider: params["provider"]}
    changeset = User.changeset(%User{}, user_params)
    insert_or_update_user(changeset)
    conn
  end

  defp insert_or_update_user(changeset) do
    case Repo.get_by(User, email: changeset.changes.email) do
      nil ->
        Repo.insert(changeset)
      user ->
        {:ok, user}
    end
  end
end
```

When I do `C-M-g` on `insert_or_update_user` inside of a `callback` function, it **should** jump to the function definition few lines below. But it does not. I see error mentioned earlier.

**Comment by jacktasia** [comment]
@mandarvaze Thanks for the example. If you're using dumb-jump within a project that is also a git repo then it'll be using `git-grep` under the hood by default.

You can get debug output to the `*Messages*` buffer by doing `(setq dumb-jump-debug t)`, but I tested your example (with an empty `.dumbjump` file as the denoter) and it worked for me so it could be an issue with `git-grep`. 

You can force using `ag` by adding `(setq dumb-jump-force-searcher 'ag)` to your dumb-jump.

**Comment by mandarvaze** [comment]
@jacktasia Here is the output from the `*Messages*` buffer after setting `(setq dumb-jump-debug t)`

```
-----
DUMB JUMP DEBUG ‘dumb-jump-run-command‘ START
----- 

cmd: 
	ag --nocolor --nogroup --elixir \\bdef\(p\)\?\\s\+insert_or_update_user\\s\*\[\ \,\\\(\]\|\\s\*insert_or_update_user\\s\*\=\[\^\=\\n\]\+\|defmodule\\s\+\(\\w\+\\.\)\*insert_or_update_user\\s\+\|defprotocol\\s\+\(\\w\+\\.\)\*insert_or_update_user\\s\+ /Users/mandar

raw results: 
	 

-----
DUMB JUMP DEBUG ‘dumb-jump-run-command‘ END
-----

-----
DUMB JUMP DEBUG ‘dumb-jump-run-command‘ (FALLBACK!) START
----- 

cmd: 
	ag --nocolor --nogroup --elixir \\binsert_or_update_user\(\?\!\[\\w-\]\) /Users/mandar

raw results: 
	 

-----
DUMB JUMP DEBUG ‘dumb-jump-run-command‘ (FALLBACK) END
-----

Took over 2s to find ’insert_or_update_user’. Please install ag or rg, or add a .dumbjump file to ’/Users/mandar’ with path exclusions

```

I compared the output with the output when it works with my python project.
Now,  I **know** the problem. 

The **last** argument is wrong. That is **not** where my code resides, and no wonder it does not find what it is looking for.
FWIW, the code where it doesn't work is NOT a git repo. Does that matter ? (I'm still learning elixir, and this is just a code from udemy course. I could put this under a git repo, if that is needed for `dumb-jump` to work.)
When I mentioned earlier that it "sometimes" works - it worked in another elixir project that is under a git repo. 
The python project where it worked, is also under a git repo.

Long story short : 
Does the project need to be under gir repo for `dumb-jump` to work ?
If not, what can I do to "fix" the last component of the `ag` command ?

**Comment by jacktasia** [comment]
@mandarvaze Thanks for the debug info. You should add an empty `.dumbjump` file to your project root. Usually dumb-jump can automatically figure out your project root by continuously looking at parent directories until it sees a [project denoter](https://github.com/jacktasia/dumb-jump/blob/497f5e68e01f71612f1d2d4040ed55e501c6ed03/dumb-jump.el#L944), but if it can't it will stop at your home directory, like in your case, and almost certainly make things slow (since it will effectively try to search _all_ your projects).

**Comment by mandarvaze** [comment]
@jacktasia  (I promise this is the last comment in this thread. Thank you so much for being patient)
**All** my projects are under `/projects`

Several of them are "git" projects and have `.git`

At the same time several of them are NOT git projects (Sometime, I'm just trying something like in the original case, I was writing almost similar code to udemy course - so I didn't think I need to commit it to git)

Instead of `.dumbjump` should I set `dumb-jump-project` to `/projects` ?

What is the difference between `dumb-jump-default-project` and `dumb-jump-project` ?

Which one is better/recommended for my setup ?

**Comment by jacktasia** [comment]
> @jacktasia (I promise this is the last comment in this thread. Thank you so much for being patient)

@mandarvaze No worries! Happy to help.

> What is the difference between dumb-jump-default-project and dumb-jump-project ?

#145 has the best breakdown on the difference.

> Which one is better/recommended for my setup ?

If your projects have their own directories then the easiest thing is just to do `touch .dumbjump` inside each directory. So dumb-jump knows the bounds of each project. This should be the fastest for both you and dumb-jump. If they don't each have their own directories then setting `dumb-jump-project` is probably your best bet, but could get slow depending how much stuff is in there, which is why I think a `.dumbjump` file is the best route. 

Hope that helps, but happy to expand on anything if you have more questions or I missed something.
